File: internal/contract/site_address_test.go
package contract

import "testing"

func TestParseSiteAddress(t *testing.T) {
	for _, tc := range []struct {
		input, host string
		port        int
	}{
		{"Example.COM", "example.com", 0}, {"Example.COM:08443", "example.com", 8443},
		{"192.168.1.10:8080", "192.168.1.10", 8080}, {"example.com:80", "example.com", 80},
		{"example.com:65535", "example.com", 65535},
	} {
		host, port, err := ParseSiteAddress(tc.input)
		if err != nil || host != tc.host || port != tc.port {
			t.Fatalf("%s: %q %d %v", tc.input, host, port, err)
		}
	}
	for _, input := range []string{"", " example.com", "example.com ", "example.com:0", "example.com:65536", "example.com:-1", "example.com:+80", "example.com:", "example.com:80/path", "https://example.com:8443", "a@b.com:80", "foo;touch:80", "../foo:80", "example..com:80", "example.com.:80", "192.168.1.1", "[::1]:8080", "x.com:80\n", "x.com:1e3", "x.com:000080"} {
		if _, _, err := ParseSiteAddress(input); err == nil {
			t.Fatalf("accepted %q", input)
		}
	}
}

File: web/src/lib/siteAddress.test.ts
import { describe, expect, it } from 'vitest'
import { parseSiteAddress, sitePublicURL } from './siteAddress'
import type { Site } from '../types/api'

describe('site address', () => {
  it('keeps domains on the legacy path and selects HTTP only with an explicit valid port', () => {
    expect(parseSiteAddress('Example.COM')).toEqual({ host: 'example.com', port: undefined })
    expect(parseSiteAddress('example.com:08443')).toEqual({ host: 'example.com', port: 8443 })
    expect(parseSiteAddress('192.168.1.10:8080')).toEqual({ host: '192.168.1.10', port: 8080 })
    for (const input of ['192.168.1.10', '999.1.1.1:80', 'x.com:0', 'x.com:65536', 'x.com:', 'x.com:+80', 'https://x.com:80', 'x.com:80/path', 'x.com:80\n', 'x.com:000080']) {
      expect(parseSiteAddress(input), input).toBeUndefined()
    }
  })
  it('uses real listeners even when HTTP has a leftover certificate and handles old API responses', () => {
    const site = { primaryDomain: 'example.com', certificate: { status: 'valid' }, accessUrls: ['http://example.com:8443'] } as Site
    expect(sitePublicURL(site)).toBe('http://example.com:8443')
    expect(sitePublicURL({ ...site, accessUrls: ['http://example.com:80', 'https://example.com:9443'] })).toBe('https://example.com:9443')
    expect(sitePublicURL({ ...site, accessUrls: undefined })).toBe('https://example.com')
    expect(sitePublicURL({ ...site, accessUrls: ['javascript:alert(1)', 'https://other.com:8443'] })).toBe('https://example.com')
    expect(sitePublicURL({ ...site, primaryDomain: '2001:db8::1', accessUrls: ['http://[2001:db8::1]:8080'] })).toBe('http://[2001:db8::1]:8080')
  })
})
